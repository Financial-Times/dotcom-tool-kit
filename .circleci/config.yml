version: 2.1

orbs:
  node: circleci/node@4.6.0
  orb-tools: circleci/orb-tools@10.0.5

references:
  #
  # Workspace
  #
  container_config_node:
    &container_config_node
    working_directory: ~/project/build
    docker:
      - image: cimg/node:<< parameters.node-version >>
    parameters:
      node-version:
        default: '16.14'
        type: string

  workspace_root: &workspace_root ~/project

  attach_workspace:
    &attach_workspace
    attach_workspace:
      at: *workspace_root

  #
  # Cache creation
  #
  create_cache:
    &create_cache
    save_cache:
      key: cache-v1-{{ .Branch }}-{{ checksum "./package.json" }}
      paths:
        - ./node_modules/

  #
  # Cache restoration
  #
  restore_cache:
    &restore_cache
    restore_cache:
      keys:
        - cache-v1-{{ .Branch }}-{{ checksum "./package.json" }}

  #
  # Filters
  #
  filters_only_main:
    &filters_only_main
    branches:
      only:
        - main

  filters_only_renovate_nori:
    &filters_only_renovate_nori
    branches:
      only: /(^renovate-.*|^nori\/.*)/

  filters_ignore_tags_renovate_nori_build_main:
    &filters_ignore_tags_renovate_nori_build_main
    tags:
      ignore: /.*/
    branches:
      ignore: /(^renovate-.*|^nori\/.*|^gh-pages|^main$)/

  filters_main_branch:
    &filters_main_branch
    branches:
      only:
        - main

  filters_release_build:
    &filters_release_build
    tags:
      only:
        - /^[\w-]+-v\d+\.\d+\.\d+$/
    branches:
      ignore: /.*/

  filters_prerelease_build:
    &filters_prerelease_build
    tags:
      only:
        - /^[\w-]+-v\d+\.\d+\.\d+(?:-\w+\.\d+)$/
    branches:
      ignore: /.*/

  filters_orbs_release_build:
    &filters_orbs_release_build
    tags:
      only:
        # CircleCI doesn't allow prerelease versions to be published
        - /^orb-v\d+\.\d+\.\d+$/
    branches:
      ignore: /.*/

jobs:
  build:
    <<: *container_config_node
    steps:
      - checkout
      - *restore_cache
      - node/install-npm:
          version: '8'
      - run:
          name: Install project dependencies
          command: npm install
      - run:
          name: Compile packages
          command: npm run build
      - *create_cache
      - persist_to_workspace:
          root: *workspace_root
          paths:
            - build

  lint:
    <<: *container_config_node
    steps:
      - *attach_workspace
      - run:
          name: Run ESLint
          command: npm run lint

  test:
    <<: *container_config_node
    steps:
      - *attach_workspace
      - run:
          name: Run package tests
          command: npm run test

  release-please:
    <<: *container_config_node
    steps:
      - *attach_workspace
      - run:
          name: Update release-please release PR
          command: npx release-please@13.17.0 release-pr --token=${RELEASE_PLEASE_GITHUB_TOKEN}
            --repo-url=${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}
      - run:
          name: Release any unreleased PR merges
          command: npx release-please@13.17.0 github-release
            --token=${RELEASE_PLEASE_GITHUB_TOKEN}
            --repo-url=${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}
            --monorepo-tags

  publish:
    <<: *container_config_node
    steps:
      - *attach_workspace
      - run:
          name: Set npm auth token
          command: echo "//registry.npmjs.org/:_authToken=${NPM_AUTH_TOKEN}" >
            ${HOME}/.npmrc
      - run:
          name: NPM publish
          command: ./scripts/circleci-publish.sh

  prepublish:
    <<: *container_config_node
    steps:
      - *attach_workspace
      - run:
          name: Set npm auth token
          command: echo "//registry.npmjs.org/:_authToken=${NPM_AUTH_TOKEN}" >
            ${HOME}/.npmrc
      - run:
          name: NPM publish
          command: ./scripts/circleci-publish.sh --tag=prerelease

workflows:
  version: 2

  build-test:
    when:
      not:
        equal:
          - scheduled_pipeline
          - << pipeline.trigger_source >>
    jobs:
      - build:
          filters:
            <<: *filters_ignore_tags_renovate_nori_build_main
          name: build-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - test:
          requires:
            - build-v<< matrix.node-version >>
          name: test-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - lint:
          requires:
            - build-v<< matrix.node-version >>
          name: lint-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]

  release-please:
    when:
      not:
        equal:
          - scheduled_pipeline
          - << pipeline.trigger_source >>
    jobs:
      - release-please:
          filters:
            <<: *filters_only_main

  renovate-nori-build-test:
    when:
      not:
        equal:
          - scheduled_pipeline
          - << pipeline.trigger_source >>
    jobs:
      - waiting-for-approval:
          type: approval
          filters:
            <<: *filters_only_renovate_nori
      - build:
          requires:
            - waiting-for-approval
          name: build-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - test:
          requires:
            - build-v<< matrix.node-version >>
          name: test-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - lint:
          requires:
            - build-v<< matrix.node-version >>
          name: lint-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]

  build-test-publish:
    when:
      not:
        equal:
          - scheduled_pipeline
          - << pipeline.trigger_source >>
    jobs:
      - build:
          filters:
            <<: *filters_release_build
          name: build-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - test:
          filters:
            <<: *filters_release_build
          requires:
            - build-v<< matrix.node-version >>
          name: test-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - lint:
          filters:
            <<: *filters_release_build
          requires:
            - build-v<< matrix.node-version >>
          name: lint-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - publish:
          context: npm-publish-token
          filters:
            <<: *filters_release_build
          requires:
            - lint-v16.14
            - test-v16.14

  build-test-prepublish:
    when:
      not:
        equal:
          - scheduled_pipeline
          - << pipeline.trigger_source >>
    jobs:
      - build:
          filters:
            <<: *filters_prerelease_build
          name: build-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - test:
          filters:
            <<: *filters_prerelease_build
          requires:
            - build-v<< matrix.node-version >>
          name: test-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - lint:
          filters:
            <<: *filters_prerelease_build
          requires:
            - build-v<< matrix.node-version >>
          name: lint-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - prepublish:
          context: npm-publish-token
          filters:
            <<: *filters_prerelease_build
          requires:
            - lint-v16.14
            - test-v16.14

  nightly:
    when:
      and:
        - equal:
            - scheduled_pipeline
            - << pipeline.trigger_source >>
        - equal:
            - nightly
            - << pipeline.schedule.name >>
    jobs:
      - build:
          context: next-nightly-build
          name: build-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]
      - test:
          requires:
            - build-v<< matrix.node-version >>
          context: next-nightly-build
          name: test-v<< matrix.node-version >>
          matrix:
            parameters:
              node-version: [ '16.14', '14.19' ]

  # Prior to producing a development orb (which requires credentials) basic validation, linting, and even unit testing can be performed.
  # This workflow will run on every commit
  orb-test-pack:
    when:
      not:
        equal:
          - scheduled_pipeline
          - << pipeline.trigger_source >>
    jobs:
      # Lint Yaml files
      - orb-tools/lint:
          lint-dir: orb
      # Pack orb source
      - orb-tools/pack:
          source-dir: orb/src
      # Publish development version(s) of the orb.
      - orb-tools/publish-dev:
          orb-name: financial-times/dotcom-tool-kit
          requires:
            - orb-tools/lint
            - orb-tools/pack

  orb-integration-test_deploy:
    when:
      not:
        equal:
          - scheduled_pipeline
          - << pipeline.trigger_source >>
    jobs:
      # Lint Yaml files
      - orb-tools/lint:
          lint-dir: orb
          filters:
            <<: *filters_orbs_release_build
      # Pack orb source
      - orb-tools/pack:
          source-dir: orb/src
          filters:
            <<: *filters_orbs_release_build
      - orb-tools/publish:
          attach-workspace: true
          # Strip leading 'orb-v' from git tag
          orb-ref: financial-times/dotcom-tool-kit@${CIRCLE_TAG:5}
          filters:
            <<: *filters_orbs_release_build
          requires:
            # - integration-test-1
            - orb-tools/lint
            - orb-tools/pack
