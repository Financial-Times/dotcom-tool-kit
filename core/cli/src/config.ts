import path from 'path'
import type { Logger } from 'winston'

import type { HookTask } from './hook'
import { loadPlugin, resolvePlugin } from './plugin'
import { Conflict, findConflicts, withoutConflicts, isConflict } from './conflict'
import { ToolKitConflictError, ToolKitError } from '@dotcom-tool-kit/error'
import { TaskClass, Hook, mapValidated, Plugin, reduceValidated, Validated } from '@dotcom-tool-kit/types'
import {
  formatTaskConflicts,
  formatUndefinedHookTasks,
  formatUnusedOptions,
  formatHookTaskConflicts,
  formatHookConflicts,
  formatOptionConflicts,
  formatUninstalledHooks,
  formatMissingTasks
} from './messages'

export interface PluginOptions {
  options: Record<string, unknown>
  plugin: Plugin
  forPlugin: Plugin
}

export interface RawConfig {
  root: string
  plugins: { [id: string]: Validated<Plugin> }
  resolvedPlugins: Set<Plugin>
  tasks: { [id: string]: TaskClass | Conflict<TaskClass> }
  hookTasks: { [id: string]: HookTask | Conflict<HookTask> }
  options: { [id: string]: PluginOptions | Conflict<PluginOptions> | undefined }
  hooks: { [id: string]: Hook<unknown> | Conflict<Hook<unknown>> }
}

export type ValidPluginsConfig = Omit<RawConfig, 'plugins'> & {
  plugins: { [id: string]: Plugin }
}

export interface ValidConfig extends ValidPluginsConfig {
  tasks: { [id: string]: TaskClass }
  hookTasks: { [id: string]: HookTask }
  options: { [id: string]: PluginOptions }
  hooks: { [id: string]: Hook<unknown> }
}

const coreRoot = path.resolve(__dirname, '../')

export const createConfig = (): RawConfig => ({
  root: coreRoot,
  plugins: {},
  resolvedPlugins: new Set(),
  tasks: {},
  hookTasks: {},
  options: {},
  hooks: {}
})

async function asyncFilter<T>(items: T[], predicate: (item: T) => Promise<boolean>): Promise<T[]> {
  const results = await Promise.all(items.map(async (item) => ({ item, keep: await predicate(item) })))

  return results.filter(({ keep }) => keep).map(({ item }) => item)
}

export function validateConfig(config: ValidPluginsConfig): asserts config is ValidConfig {
  const hookTaskConflicts = findConflicts(Object.values(config.hookTasks))
  const hookConflicts = findConflicts(Object.values(config.hooks))
  const taskConflicts = findConflicts(Object.values(config.tasks))
  const optionConflicts = findConflicts(Object.values(config.options))

  const definedHookTaskConflicts = hookTaskConflicts.filter((conflict) => {
    return conflict.conflicting[0].id in config.hooks
  })

  let shouldThrow = false
  const error = new ToolKitConflictError(
    'There are problems with your Tool Kit configuration.',
    hookTaskConflicts.map((conflict) => ({
      hook: conflict.conflicting[0].id,
      conflictingTasks: conflict.conflicting.flatMap((hook) =>
        hook.tasks.map((task) => ({ task, plugin: hook.plugin.id }))
      )
    }))
  )
  error.details = ''

  if (
    hookConflicts.length > 0 ||
    definedHookTaskConflicts.length > 0 ||
    taskConflicts.length > 0 ||
    optionConflicts.length > 0
  ) {
    shouldThrow = true

    if (hookConflicts.length) {
      error.details += formatHookConflicts(hookConflicts)
    }

    if (definedHookTaskConflicts.length) {
      error.details += formatHookTaskConflicts(definedHookTaskConflicts)
    }

    if (taskConflicts.length) {
      error.details += formatTaskConflicts(taskConflicts)
    }

    if (optionConflicts.length) {
      error.details += formatOptionConflicts(optionConflicts)
    }
  }

  const configuredHookTasks = withoutConflicts(Object.values(config.hookTasks))
  const definedHookIds = new Set(Object.keys(config.hooks))
  const undefinedHookTasks = configuredHookTasks.filter((hookTask) => {
    // we only care about undefined hooks that were configured by the app, not default config from plugins
    const fromApp = hookTask.plugin.root === process.cwd()
    const hookDefined = definedHookIds.has(hookTask.id)
    return fromApp && !hookDefined
  })

  if (undefinedHookTasks.length > 0) {
    shouldThrow = true
    error.details += formatUndefinedHookTasks(undefinedHookTasks, Array.from(definedHookIds))
  }

  const unusedOptions = Object.entries(config.options)
    .filter(
      ([, option]) =>
        option && !isConflict(option) && !option.forPlugin && option.plugin.root === process.cwd()
    )
    .map(([id]) => id)
  if (unusedOptions.length > 0) {
    shouldThrow = true
    error.details += formatUnusedOptions(unusedOptions, Object.keys(config.plugins))
  }

  const missingTasks = configuredHookTasks
    .map((hook) => ({
      hook,
      tasks: hook.tasks.filter((id) => !config.tasks[id])
    }))
    .filter(({ tasks }) => tasks.length > 0)

  if (missingTasks.length > 0) {
    shouldThrow = true
    error.details += formatMissingTasks(missingTasks, Object.keys(config.tasks))
  }

  if (shouldThrow) {
    throw error
  }
}

export function validatePlugins(config: RawConfig): Validated<ValidPluginsConfig> {
  const validatedPlugins = reduceValidated(
    Object.entries(config.plugins).map(([id, plugin]) => mapValidated(plugin, (p) => [id, p] as const))
  )
  return mapValidated(validatedPlugins, (plugins) => ({ ...config, plugins: Object.fromEntries(plugins) }))
}

export async function checkInstall(config: ValidConfig): Promise<void> {
  const definedHooks = withoutConflicts(Object.values(config.hooks))
  const uninstalledHooks = await asyncFilter(definedHooks, async (hook) => {
    return !(await hook.check())
  })

  if (uninstalledHooks.length > 0) {
    const error = new ToolKitError('There are problems with your Tool Kit installation.')
    error.details = formatUninstalledHooks(uninstalledHooks)
    throw error
  }
}

export function loadConfig(logger: Logger, options?: { validate?: true }): Promise<ValidConfig>
export function loadConfig(logger: Logger, options?: { validate?: false }): Promise<RawConfig>

export async function loadConfig(logger: Logger, { validate = true } = {}): Promise<ValidConfig | RawConfig> {
  const config = createConfig()

  // start loading config and child plugins, starting from the consumer app directory
  const rootPlugin = await loadPlugin('app root', config, logger)
  if (!rootPlugin.valid) {
    const error = new ToolKitError('root plugin was not valid!')
    error.details = rootPlugin.reasons.join('\n\n')
    throw error
  }
  const validRootPlugin = rootPlugin.value

  const validatedPluginConfig = validatePlugins(config)

  if (!validatedPluginConfig.valid) {
    const error = new ToolKitError('config was not valid!')
    error.details = validatedPluginConfig.reasons.join('\n\n')
    throw error
  }
  const validPluginConfig = validatedPluginConfig.value

  // collate root plugin and descendent hooks, options etc into config
  resolvePlugin(validRootPlugin, validPluginConfig, logger)

  if (validate) {
    validateConfig(validPluginConfig)
    return validPluginConfig
  }

  return config
}
