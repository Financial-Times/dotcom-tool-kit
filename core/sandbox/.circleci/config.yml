# CONFIG GENERATED BY DOTCOM-TOOL-KIT, DO NOT EDIT BY HAND
version: 2.1
orbs:
  tool-kit: financial-times/dotcom-tool-kit@dev:alpha
jobs:
  checkout:
    docker:
      - image: cimg/base:stable
    steps:
      - checkout
      - tool-kit/persist-workspace:
          path: .
workflows:
  version: 2
  tool-kit:
    jobs:
      - checkout
      - waiting-for-approval:
          type: approval
          filters:
            branches:
              only: /(^renovate-.*|^nori/.*)/
      - tool-kit/setup:
          requires:
            - checkout
            - waiting-for-approval
      - tool-kit/build:
          requires:
            - tool-kit/setup
      - tool-kit/test:
          requires:
            - tool-kit/build
      - tool-kit/heroku-provision:
          requires:
            - tool-kit/setup
          filters:
            branches:
              ignore: /(^renovate-.*|^nori/.*|^main)/
      - tool-kit/heroku-staging:
          requires:
            - tool-kit/setup
          filters:
            branches:
              only: main
      - tool-kit/e2e-test-review:
          requires:
            - tool-kit/heroku-provision
      - tool-kit/e2e-test-staging:
          requires:
            - tool-kit/heroku-staging
      - tool-kit/heroku-promote:
          requires:
            - tool-kit/test
            - tool-kit/e2e-test-staging
          filters:
            branches:
              only: main
      # if repo is a component instead of app, use publish instead of heroku jobs
      # - tool-kit/publish:
      #     requires:
      #       - tool-kit/test
      #     filters:
      #       branches:
      #         ignore: /.*/
      #       tags:
      #         only: /^v\d+\.\d+\.\d+(-.+)?/
  nightly:
    triggers:
      - schedule:
          cron: 0 0 * * *
          filters:
            branches:
              only: main
    jobs:
      - tool-kit/setup
      - tool-kit/build:
          requires:
            - tool-kit/setup
      - tool-kit/test:
          requires:
            - tool-kit/build
