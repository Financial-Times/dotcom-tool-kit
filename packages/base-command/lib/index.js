"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
tslib_1.__exportStar(require("@oclif/command"), exports);
class BaseCommand extends command_1.Command {
}
exports.default = BaseCommand;
