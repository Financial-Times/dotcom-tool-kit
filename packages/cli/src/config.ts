import path from 'path'

import type { TaskClass } from '@dotcom-tool-kit/task'
import type { HookTask, HookClass } from './hook'
import { loadPluginConfig, Plugin } from './plugin'
import { Conflict, findConflicts, withoutConflicts, isConflict } from './conflict'
import { ToolKitConflictError, ToolKitError } from '@dotcom-tool-kit/error'
import {
  formatTaskConflicts,
  formatUndefinedHookTasks,
  formatUnusedOptions,
  formatHookTaskConflicts,
  formatHookConflicts,
  formatOptionConflicts,
  formatUninstalledHooks,
  formatMissingTasks
} from './messages'

export interface PluginOptions {
  options: Record<string, unknown>
  plugin: Plugin
  forPlugin: Plugin
}

export interface Config {
  root: string
  plugins: { [id: string]: Plugin }
  tasks: { [id: string]: TaskClass | Conflict<TaskClass> }
  hookTasks: { [id: string]: HookTask | Conflict<HookTask> }
  options: { [id: string]: PluginOptions | Conflict<PluginOptions> | undefined }
  hooks: { [id: string]: HookClass | Conflict<HookClass> }
}

export interface ValidConfig extends Config {
  tasks: { [id: string]: TaskClass }
  hookTasks: { [id: string]: HookTask }
  options: { [id: string]: PluginOptions }
  hooks: { [id: string]: HookClass }
}

const coreRoot = path.resolve(__dirname, '../')

const createConfig = (): Config => ({
  root: coreRoot,
  plugins: {},
  tasks: {},
  hookTasks: {},
  options: {},
  hooks: {}
})

async function asyncFilter<T>(items: T[], predicate: (item: T) => Promise<boolean>): Promise<T[]> {
  const results = await Promise.all(items.map(async (item) => ({ item, keep: await predicate(item) })))

  return results.filter(({ keep }) => keep).map(({ item }) => item)
}

export async function validateConfig(config: Config): Promise<ValidConfig> {
  const hookTaskConflicts = findConflicts(Object.values(config.hookTasks))
  const hookConflicts = findConflicts(Object.values(config.hooks))
  const taskConflicts = findConflicts(Object.values(config.tasks))
  const optionConflicts = findConflicts(Object.values(config.options))

  const definedHookTaskConflicts = hookTaskConflicts.filter((conflict) => {
    return conflict.conflicting[0].id in config.hooks
  })

  let shouldThrow = false
  const error = new ToolKitConflictError(
    'There are problems with your Tool Kit configuration.',
    hookTaskConflicts.map((conflict) => ({
      hook: conflict.conflicting[0].id,
      conflictingTasks: conflict.conflicting.flatMap((hook) =>
        hook.tasks.map((task) => ({ task, plugin: hook.plugin.id }))
      )
    }))
  )
  error.details = ''

  if (
    hookConflicts.length > 0 ||
    definedHookTaskConflicts.length > 0 ||
    taskConflicts.length > 0 ||
    optionConflicts.length > 0
  ) {
    shouldThrow = true

    if (hookConflicts.length) {
      error.details += formatHookConflicts(hookConflicts)
    }

    if (definedHookTaskConflicts.length) {
      error.details += formatHookTaskConflicts(definedHookTaskConflicts)
    }

    if (taskConflicts.length) {
      error.details += formatTaskConflicts(taskConflicts)
    }

    if (optionConflicts.length) {
      error.details += formatOptionConflicts(optionConflicts)
    }
  }

  const configuredHookTasks = withoutConflicts(Object.values(config.hookTasks))
  const definedHookIds = new Set(Object.keys(config.hooks))
  const undefinedHookTasks = configuredHookTasks.filter((hookTask) => {
    // we only care about undefined hooks that were configured by the app, not default config from plugins
    const fromApp = hookTask.plugin.root === process.cwd()
    const hookDefined = definedHookIds.has(hookTask.id)
    return fromApp && !hookDefined
  })

  if (undefinedHookTasks.length > 0) {
    shouldThrow = true
    error.details += formatUndefinedHookTasks(undefinedHookTasks, Array.from(definedHookIds))
  }

  const unusedOptions = Object.entries(config.options)
    .filter(
      ([, option]) =>
        option && !isConflict(option) && !option.forPlugin && option.plugin.root === process.cwd()
    )
    .map(([id]) => id)
  if (unusedOptions.length > 0) {
    shouldThrow = true
    error.details += formatUnusedOptions(unusedOptions, Object.keys(config.plugins))
  }

  const missingTasks = configuredHookTasks
    .map((hook) => ({
      hook,
      tasks: hook.tasks.filter((id) => !config.tasks[id])
    }))
    .filter(({ tasks }) => tasks.length > 0)

  if (missingTasks.length > 0) {
    shouldThrow = true
    error.details += formatMissingTasks(missingTasks, Object.keys(config.tasks))
  }

  if (shouldThrow) {
    throw error
  }

  return config as ValidConfig
}

export async function checkInstall(config: ValidConfig): Promise<void> {
  const definedHooks = withoutConflicts(Object.values(config.hooks))
  const uninstalledHooks = await asyncFilter(definedHooks, async (Hook) => {
    const hook = new Hook()
    return !(await hook.check())
  })

  if (uninstalledHooks.length > 0) {
    const error = new ToolKitError('There are problems with your Tool Kit installation.')
    error.details = formatUninstalledHooks(uninstalledHooks)
    throw error
  }
}

export function loadConfig(options?: { validate?: true }): Promise<ValidConfig>
export function loadConfig(options?: { validate?: false }): Promise<Config>

export async function loadConfig({ validate = true } = {}): Promise<ValidConfig | Config> {
  // start loading config and child plugins, starting from the consumer app directory
  const config = await loadPluginConfig(
    {
      id: 'app root',
      root: process.cwd()
    },
    createConfig()
  )

  return validate ? validateConfig(config) : config
}
