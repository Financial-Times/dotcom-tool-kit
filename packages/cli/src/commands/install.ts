import { isConflict } from '../conflict'
import { config } from '../config'
import { Task } from '@dotcom-tool-kit/task'
import { ToolKitError } from '@dotcom-tool-kit/error'

const isRejected = (result: PromiseSettledResult<unknown>): result is PromiseRejectedResult =>
  result.status === 'rejected'

export default class InstallCommand extends Task {
  static description = 'run lifecycle commands'

  async run(): Promise<void> {
    const results = await Promise.allSettled(
      Object.values(config.lifecycles).map(async (Lifecycle) => {
        if (isConflict(Lifecycle)) return

        const lifecycle = new Lifecycle()

        if (!(await lifecycle.check())) {
          await lifecycle.install()
        }
      })
    )

    const errors = results.filter(isRejected).map((result) => result.reason as Error)

    if (errors.length) {
      const error = new ToolKitError('could not automatically install lifecycles:')
      error.details = errors.map((error) => error.message).join('\n\n')
      throw error
    }
  }
}
