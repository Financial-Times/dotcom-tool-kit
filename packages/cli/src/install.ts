import { ToolKitError } from '@dotcom-tool-kit/error'
import { Config, loadConfig } from './config'

export default async function installHooks(): Promise<Config> {
  const config = await loadConfig({ checkInstall: false })

  const tasks = Object.values(config.hooks).map((Hook) => async () => {
    const hook = new Hook()

    if (!(await hook.check())) {
      await hook.install()
    }
  })

  const errors: Error[] = []
  for (const task of tasks) {
    try {
      await task()
    } catch (err) {
      if (err instanceof Error) {
        errors.push(err)
      } else {
        throw err
      }
    }
  }

  if (errors.length) {
    const error = new ToolKitError('could not automatically install hooks:')
    error.details = errors.map((error) => error.message).join('\n\n')
    throw error
  }

  return config
}
