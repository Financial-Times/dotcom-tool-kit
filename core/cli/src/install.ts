import { z } from 'zod'
import { ToolKitError } from '@dotcom-tool-kit/error'
import { OptionKey, setOptions } from '@dotcom-tool-kit/options'
import groupBy from 'lodash/groupBy'
import type { Logger } from 'winston'
import { loadConfig } from './config'
import { hasConfigChanged, updateHashes } from './config/hash'
import {
  Hook,
  HookClass,
  ValidConfig,
  Validated,
  invalid,
  reduceValidated,
  valid
} from '@dotcom-tool-kit/types'
import { reducePluginHookInstallations } from './plugin/reduce-installations'
import { findConflicts, withoutConflicts } from '@dotcom-tool-kit/types/src/conflict'
import { Options as HookSchemaOptions, HookSchemas } from '@dotcom-tool-kit/types/lib/hooks'
import { formatUninstalledHooks } from './messages'
import { importEntryPoint } from './plugin/entry-point'
import { runInit } from './init'

// implementation of the Array#some method that supports asynchronous predicates
async function asyncSome<T>(arr: T[], pred: (x: T) => Promise<boolean>): Promise<boolean> {
  for (const val of arr) {
    if (await pred(val)) {
      return true
    }
  }
  return false
}

// implementation of the Array#filter method that supports asynchronous predicates
async function asyncFilter<T>(items: T[], predicate: (item: T) => Promise<boolean>): Promise<T[]> {
  const results = await Promise.all(items.map(async (item) => ({ item, keep: await predicate(item) })))

  return results.filter(({ keep }) => keep).map(({ item }) => item)
}

const loadHookEntrypoints = async (
  logger: Logger,
  config: ValidConfig
): Promise<Validated<Record<string, HookClass>>> => {
  const hookResultEntries = reduceValidated(
    await Promise.all(
      Object.entries(config.hooks).map(async ([hookName, entryPoint]) => {
        const hookResult = await importEntryPoint(Hook, entryPoint)
        return hookResult.map((hookClass) => [hookName, hookClass as HookClass] as const)
      })
    )
  )

  return hookResultEntries.map((hookEntries) => Object.fromEntries(hookEntries))
}

export const loadHookInstallations = async (
  logger: Logger,
  config: ValidConfig
): Promise<Validated<Hook<z.ZodType, unknown>[]>> => {
  const hookClassResults = await loadHookEntrypoints(logger, config)
  const installationResults = await hookClassResults
    .map((hookClasses) =>
      reducePluginHookInstallations(logger, config, hookClasses, config.plugins['app root'])
    )
    .awaitValue()

  const installationsWithoutConflicts = installationResults.flatMap((installations) => {
    const conflicts = findConflicts(installations)

    if (conflicts.length) {
      return invalid<[]>([])
    }

    return valid(withoutConflicts(installations))
  })

  return installationsWithoutConflicts.map((installations) => {
    return installations.map(({ hookConstructor, forHook, options }) => {
      const schema = HookSchemas[forHook as keyof HookSchemaOptions]
      const parsedOptions = schema ? schema.parse(options) : {}
      return new hookConstructor(logger, forHook, parsedOptions)
    })
  })
}

export async function checkInstall(logger: Logger, config: ValidConfig): Promise<void> {
  if (!(await hasConfigChanged(logger))) {
    return
  }

  const hooks = (await loadHookInstallations(logger, config)).unwrap('hooks are invalid')

  const uninstalledHooks = await asyncFilter(hooks, async (hook) => {
    return !(await hook.isInstalled())
  })

  if (uninstalledHooks.length > 0) {
    const error = new ToolKitError('There are problems with your Tool Kit installation.')
    error.details = formatUninstalledHooks(uninstalledHooks)
    throw error
  }

  await updateHashes()
}

export default async function installHooks(logger: Logger): Promise<ValidConfig> {
  const config = await loadConfig(logger)

  for (const pluginOptions of Object.values(config.options)) {
    if (pluginOptions.forPlugin) {
      setOptions(pluginOptions.forPlugin.id as OptionKey, pluginOptions.options)
    }
  }

  await runInit(logger, config)

  const errors: Error[] = []
  const hooks = (await loadHookInstallations(logger, config)).unwrap('hooks are invalid')

  // group hooks without an installGroup separately so that their check()
  // method runs independently
  const groups = groupBy(hooks, (hook) => hook.installGroup ?? '__' + hook.id)

  for (const group of Object.values(groups)) {
    try {
      if (await asyncSome(group, async (hook) => !(await hook.isInstalled()))) {
        let state = undefined
        for (const hook of group) {
          state = await hook.install(state)
        }
        if (state) {
          try {
            await group[0].commitInstall(state)
          } catch (err) {
            if (err instanceof Error) {
              errors.push(err)
            } else {
              throw err
            }
          }
        }
      }
    } catch (err) {
      if (err instanceof Error) {
        errors.push(err)
      } else {
        throw err
      }
    }
  }

  if (errors.length) {
    const error = new ToolKitError('could not automatically install hooks:')
    error.details = errors.map((error) => `- ${error.message}`).join('\n')
    throw error
  }

  await updateHashes()

  return config
}
