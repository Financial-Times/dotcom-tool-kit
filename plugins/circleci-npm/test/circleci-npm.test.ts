import * as YAML from 'yaml'
import path from 'path'
import CircleCi from '@dotcom-tool-kit/circleci/lib/circleci-config'
import winston, { Logger } from 'winston'
import { loadConfig } from 'dotcom-tool-kit/lib/config'
import { loadHookInstallations } from 'dotcom-tool-kit/lib/install'

const logger = winston as unknown as Logger

describe('circleci-npm', () => {
  describe('config integration test', () => {
    it('should generate a .circleci/config.yml with the base config from circleci-npm/.toolkitrc.yml', async () => {
      const config = await loadConfig(logger, { root: path.resolve(__dirname, '..') })
      const hookInstallationsPromise = loadHookInstallations(logger, config).then((validated) =>
        validated.unwrap('hooks were invalid')
      )

      await expect(hookInstallationsPromise).resolves.toEqual([expect.any(CircleCi)])

      const installation = (await hookInstallationsPromise)[0]

      expect(YAML.stringify(await installation.install())).toMatchSnapshot()
    })
  })
})
