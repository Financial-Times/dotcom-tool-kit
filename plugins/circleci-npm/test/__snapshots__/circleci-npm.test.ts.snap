// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`circleci-npm config integration test should generate a .circleci/config.yml with the base config from circleci-npm/.toolkitrc.yml 1`] = `
"version: 2.1
orbs:
  tool-kit: financial-times/dotcom-tool-kit@5
executors:
  node:
    docker:
      - image: cimg/node:18.19-browsers
jobs:
  checkout:
    docker:
      - image: cimg/base:stable
    steps:
      - checkout
      - persist_to_workspace:
          root: .
          paths:
            - .
workflows:
  tool-kit:
    when:
      not:
        equal:
          - scheduled_pipeline
          - << pipeline.trigger_source >>
    jobs:
      - checkout:
          filters:
            tags:
              only: /^v\\d+\\.\\d+\\.\\d+(-.+)?/
      - tool-kit/setup:
          executor: node
          requires:
            - checkout
          filters:
            tags:
              only: /^v\\d+\\.\\d+\\.\\d+(-.+)?/
      - tool-kit/build:
          executor: node
          requires:
            - tool-kit/setup
      - tool-kit/test:
          executor: node
          requires:
            - tool-kit/build
      - tool-kit/publish-tag:
          executor: node
          requires:
            - tool-kit/test
          context: npm-publish-token
          filters:
            branches:
              ignore: /.*/
  nightly:
    when:
      and:
        - equal:
            - scheduled_pipeline
            - << pipeline.trigger_source >>
        - equal:
            - nightly
            - << pipeline.schedule.name >>
    jobs:
      - checkout
      - tool-kit/setup:
          executor: node
          requires:
            - checkout
      - tool-kit/build:
          executor: node
          requires:
            - tool-kit/setup
      - tool-kit/test:
          executor: node
          requires:
            - tool-kit/build
"
`;
